lib/ruby_lsp/handler.rb

```diff
@@ -1,118 +1,157 @@
 # typed: strict
 # frozen_string_literal: true
 
-require "ruby_lsp/requests"
-require "ruby_lsp/store"
-require "ruby_lsp/queue"
-
 module RubyLsp
   Interface = LanguageServer::Protocol::Interface
   Constant = LanguageServer::Protocol::Constant
   Transport = LanguageServer::Protocol::Transport
 
   class Handler
     extend T::Sig
-    VOID = T.let(Object.new.freeze, Object)
-
-    class RequestHandler < T::Struct
-      extend T::Sig
-
-      const :action, T.proc.params(request: T::Hash[Symbol, T.untyped]).returns(T.untyped)
-      const :parallel, T::Boolean
-      prop :error_handler,
-        T.nilable(T.proc.params(error: Exception, request: T::Hash[Symbol, T.untyped]).void)
-
-      # A proc that runs in case a request has errored. Receives the error and the original request as arguments. Useful
-      # for displaying window messages on errors
-      sig do
-        params(
-          block: T.proc.bind(Handler).params(error: Exception, request: T::Hash[Symbol, T.untyped]).void,
-        ).void
-      end
-      def on_error(&block)
-        self.error_handler = block
-      end
-    end
-
-    class << self
-      extend T::Sig
-
-      sig { params(blk: T.proc.bind(Handler).params(arg0: T.untyped).void).void }
-      def start(&blk)
-        handler = new
-        handler.instance_exec(&blk)
-        handler.start
-      end
-    end
-
-    sig { returns(Store) }
-    attr_reader :store
 
     sig { void }
     def initialize
       @writer = T.let(Transport::Stdio::Writer.new, Transport::Stdio::Writer)
       @reader = T.let(Transport::Stdio::Reader.new, Transport::Stdio::Reader)
-      @handlers = T.let({}, T::Hash[String, RequestHandler])
       @store = T.let(Store.new, Store)
-      @queue = T.let(Queue.new(@writer, @handlers), Queue)
+
+      # The job queue is the actual list of requests we have to process
+      @job_queue = T.let(Thread::Queue.new, Thread::Queue)
+      # The jobs hash is just a way of keeping a handle to jobs based on the request ID, so we can cancel them
+      @jobs = T.let({}, T::Hash[T.any(String, Integer), Job])
+      @mutex = T.let(Mutex.new, Mutex)
+      @worker = T.let(new_worker, Thread)
```

Comment:
Does the worker get utilized at all?

Reply:
It's only used during `shutdown`. We close the `Thread::Queue`, which allows the worker thread to finish its infinite loop and then we do `@worker.join` just to wait until it's finalized.